x-pack/plugins/security/server/lib/authentication/providers/token.js

```diff
@@ -0,0 +1,215 @@
+      }
+    }
+
+    // finally, if authentication still can not be handled for this
+    // request/state combination, redirect to the login page if appropriate
+    if (authenticationResult.notHandled() && canRedirectRequest(request)) {
+      const nextURL = encodeURIComponent(`${this._options.basePath}${request.url.path}`);
+      return AuthenticationResult.redirectTo(
+        `${this._options.basePath}/login?next=${nextURL}`
+      );
+    }
+
+    return authenticationResult;
+  }
+
+  /**
+   * Redirects user to the login page preserving query string parameters.
+   * @param {Hapi.Request} request HapiJS request instance.
+   * @returns {Promise.<DeauthenticationResult>}
+   */
+  async deauthenticate(request) {
+    // Query string may contain the path where logout has been called or
+    // logout reason that login page may need to know.
+    return DeauthenticationResult.redirectTo(
+      `${this._options.basePath}/login${request.url.search}`
+    );
+  }
+
+  /**
+   * Validates whether request contains `Basic ***` Authorization header and just passes it
+   * forward to Elasticsearch backend.
+   * @param {Hapi.Request} request HapiJS request instance.
+   * @returns {Promise.<AuthenticationResult>}
+   * @private
+   */
+  async _authenticateViaHeader(request) {
+    this._options.log(['debug', 'security', 'token'], 'Trying to authenticate via header.');
+
+    const authorization = request.headers.authorization;
+    if (!authorization) {
+      this._options.log(['debug', 'security', 'token'], 'Authorization header is not presented.');
+      return AuthenticationResult.notHandled();
+    }
+
+    const authenticationSchema = authorization.split(/\s+/)[0];
+    if (authenticationSchema.toLowerCase() !== 'basic') {
+      this._options.log(['debug', 'security', 'token'], `Unsupported authentication schema: ${authenticationSchema}`);
+
+      // It's essential that we fail if non-empty, but unsupported authentication schema
+      // is provided to allow authenticator to consult other authentication providers
+      // that may support that schema.
+      return AuthenticationResult.failed(
+        Boom.badRequest(`Unsupported authentication schema: ${authenticationSchema}`)
+      );
+    }
+
+    try {
+      const user = await this._options.client.callWithRequest(request, 'shield.authenticate');
+
+      this._options.log(['debug', 'security', 'token'], 'Request has been authenticated via header.');
+
+      return AuthenticationResult.succeeded(user, { authorization });
+    } catch(err) {
+      this._options.log(['debug', 'security', 'token'], `Failed to authenticate request via header: ${err.message}`);
+      return AuthenticationResult.failed(err);
+    }
+  }
+
+  /**
+   * @param {Hapi.Request} request HapiJS request instance.
+   * @returns {Promise.<AuthenticationResult>}
+   * @private
+   */
+  async _authenticateViaLoginAttempt(request) {
```

Comment:
This is the key function that is invoked when login credentials exist for a request, which it retrieves from the `loginAttempt` object on the request.

The basic provider would be updated to use this mechanism as well. The only real difference between this function in the token provider and the basic provider is the interaction with Elasticsearch using those credentials.